cmd/attestation-service/main.go

```diff
@@ -0,0 +1,201 @@
+		if err := quoteprovider.FetchCertificates(uint(cfg.Vmpl)); err != nil {
+			logger.Error(fmt.Sprintf("failed to fetch certificates: %s", err))
+			exitCode = 1
+			return
+		}
+	}
+
+	// Remove existing socket if it exists
+	if _, err := os.Stat(socketPath); err == nil {
+		if err := os.Remove(socketPath); err != nil {
+			logger.Error(fmt.Sprintf("failed to remove existing socket: %s", err))
+			exitCode = 1
+			return
+		}
+	}
+
+	dir := socketPath[:len(socketPath)-len("/attestation.sock")]
+	if err := os.MkdirAll(dir, 0o755); err != nil {
+		logger.Error(fmt.Sprintf("failed to create socket directory: %s", err))
+		exitCode = 1
+		return
+	}
+
+	lis, err := net.Listen("unix", socketPath)
+	if err != nil {
+		logger.Error(fmt.Sprintf("failed to listen on socket: %s", err))
+		exitCode = 1
+		return
+	}
+
+	if err := os.Chmod(socketPath, 0o777); err != nil {
+		logger.Error(fmt.Sprintf("failed to chmod socket: %s", err))
+		exitCode = 1
+		return
+	}
+
+	grpcServer := grpc.NewServer()
+	svc := &service{
+		provider: provider,
+		logger:   logger,
+	}
+	attestationpb.RegisterAttestationServiceServer(grpcServer, svc)
+
+	g.Go(func() error {
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, syscall.SIGINT, syscall.SIGTERM)
+		defer signal.Stop(ch)
+
+		select {
+		case <-ch:
+			logger.Info("Received signal, shutting down...")
+			cancel()
+			grpcServer.GracefulStop()
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	})
+
+	g.Go(func() error {
+		logger.Info(fmt.Sprintf("%s started on %s", svcName, socketPath))
+		return grpcServer.Serve(lis)
+	})
+
+	if err := g.Wait(); err != nil {
+		logger.Error(fmt.Sprintf("%s terminated: %s", svcName, err))
+	}
+}
+
+type service struct {
+	attestationpb.UnimplementedAttestationServiceServer
+	provider attestation.Provider
+	logger   *slog.Logger
+}
+
+func (s *service) GetAttestation(ctx context.Context, req *attestationpb.AttestationRequest) (*attestationpb.AttestationResponse, error) {
+	var quote []byte
+	var err error
+
+	switch req.PlatformType {
+	case attestationpb.PlatformType_PLATFORM_TYPE_SNP, attestationpb.PlatformType_PLATFORM_TYPE_TDX:
+		var reportData [64]byte
+		copy(reportData[:], req.ReportData)
+		quote, err = s.provider.TeeAttestation(reportData[:])
+	case attestationpb.PlatformType_PLATFORM_TYPE_VTPM:
+		var nonce [32]byte
+		copy(nonce[:], req.Nonce)
+		quote, err = s.provider.VTpmAttestation(nonce[:])
+	case attestationpb.PlatformType_PLATFORM_TYPE_SNP_VTPM:
+		var reportData [64]byte
+		copy(reportData[:], req.ReportData)
+		var nonce [32]byte
+		copy(nonce[:], req.Nonce)
+		quote, err = s.provider.Attestation(reportData[:], nonce[:])
+	default:
+		return nil, fmt.Errorf("unsupported platform type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &attestationpb.AttestationResponse{Quote: quote}, nil
+}
+
+func (s *service) GetAzureToken(ctx context.Context, req *attestationpb.AzureTokenRequest) (*attestationpb.AzureTokenResponse, error) {
```

Comment:
Why do we need this? Is it specific for Azure flow? If this is common for all the cloud providers, can we make the API generic and let service decide which token to return?

Reply:
It is specific to azure